src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,856 @@
+#include "jfr/utilities/jfrTypes.hpp"
+#include "classfile/javaThreadStatus.hpp"
+#include "jfr/recorder/jfrRecorder.hpp"
+#include "jfr/periodic/sampling/jfrCallTrace.hpp"
+#include "jfr/recorder/storage/jfrBuffer.hpp"
+#include "jfr/utilities/jfrTime.hpp"
+#include "jfrfiles/jfrEventClasses.hpp"
+#include "runtime/mutexLocker.hpp"
+#include "runtime/threadSMR.hpp"
+#include "runtime/threadCrashProtection.hpp"
+#include "runtime/osThread.hpp"
+
+#if defined(LINUX)
+#include "signals_posix.hpp"
+
+enum JfrSampleType {
+  // no sample, because thread not in walkable state
+  NO_SAMPLE = 0,
+  // sample from thread while in Java
+  JAVA_SAMPLE = 1,
+  // sample from thread while in native
+  NATIVE_SAMPLE = 2
+};
+
+static bool thread_state_in_java(JavaThread* thread) {
+  assert(thread != nullptr, "invariant");
+  switch(thread->thread_state()) {
+    case _thread_new:
+    case _thread_uninitialized:
+    case _thread_new_trans:
+    case _thread_in_vm_trans:
+    case _thread_blocked_trans:
+    case _thread_in_native_trans:
+    case _thread_blocked:
+    case _thread_in_vm:
+    case _thread_in_native:
+      break;
+    case _thread_in_Java_trans:
+    case _thread_in_Java:
+      return true;
+    default:
+      ShouldNotReachHere();
+      break;
+  }
+  return false;
+}
+
+static bool thread_state_in_native(JavaThread* thread) {
+  assert(thread != nullptr, "invariant");
+  switch(thread->thread_state()) {
+    case _thread_new:
+    case _thread_uninitialized:
+    case _thread_new_trans:
+    case _thread_in_Java_trans:
+    case _thread_in_Java:
+          break;
+    case _thread_blocked_trans:
+    case _thread_in_vm_trans:
+    case _thread_in_native_trans:
+    case _thread_blocked:
+    case _thread_in_native:
+    case _thread_in_vm: // walking in vm causes weird bugs (assertions in G1 fail), so don't
+      return true;
+    default:
+      ShouldNotReachHere();
+      break;
+  }
+  return false;
+}
+
+static bool is_excluded(JavaThread* thread) {
+  return thread->is_hidden_from_external_view() ||
+    (os::is_readable_pointer(thread->jfr_thread_local()) && thread->jfr_thread_local()->is_excluded());
+}
+
+static JavaThread* get_java_thread_if_valid() {
+  Thread* raw_thread = Thread::current_or_null_safe();
+  JavaThread* jt;
+
+  if (raw_thread == nullptr || !raw_thread->is_Java_thread() ||
+      (jt = JavaThread::cast(raw_thread))->is_exiting()) {
+    return nullptr;
+  }
+
+  if (is_excluded(jt)) {
+    return nullptr;
+  }
+  return jt;
+}
+
+// A trace of stack frames, contains all information
+// collected in the signal handler, required to create
+// a JFR event with a stack trace
+class JfrCPUTimeTrace {
+  friend class JfrTraceQueue;
+  u4 _index;
+  JfrAsyncStackFrame* _frames;
+  JfrAsyncStackTrace _stacktrace;
+  u4 _max_frames;
+  // error code for the trace, 0 if no error
+  u4 _error;
+
+  JfrSampleType _type;
+  JfrTicks _start_time;
+  JfrTicks _end_time;
+  JavaThread* _sampled_thread;
```

Comment:
You can store the thread's traceid (u8) here instead of the Thread* itself.

Then, later, you can skip using both JFRRecordSampledThreadCallback and crash protection.

Reply:
No, we can't. This crashes the sampler. The problem is, that obtaining the traceid via `JfrThreadLocal::thread_id(jt)` acquires a lock (and more) when the traceid is not set already. So we can't use it in the signal handler.

Reply:
Hmm. That's not good. We could eliminate the spinlock, but that will require a thread assigning its id early in thread_state_new to prevent our old sampler from hitting it (and also the new one from skipping new threads). But complexity arises in that it involves reading the tid from a Java oop; so it eventually needs to be in a thread in the VM.

Have to think some more about this. Too bad you need to do this complexity.

Reply:
I have created this to help reduce the amount of crash protection you need to use: https://github.com/openjdk/jdk/pull/21756

Reply:
I'm happy to use this in the PR once it is merged.